### DISCUSSION TITLE
How can I keep polling an API endpoint for long time?

### DISCUSSION THREAD

**Discussion by offchan42** [vercel/community#6001] [Feature Question] [answered]
### Question

Suppose I have deployed an AI model to a GPU hosting service.
Typically I would request a model like this:
```js
const axios = require('axios');
const os = require('os');

const BASETEN_API_KEY = process.env.BASETEN_API_KEY;
const modelId = ''; // Replace with your model id

const data = {
  prompt: 'Extremely detailed and intricate scene of baby phoenix hatchling cuddled up resting on a pile of ashes surrounded by fire and smoke, rays of sunlight shine on the phoenix, in the background is a dense dark forest, settings: f/8 aperture, full shot, hyper realistic, 4k',
  negative_prompt: 'worst quality, low quality',
  width: 1248,
  height: 832,
  num_inference_steps: 35,
  guidance_scale: 14,
};

// Call model endpoint
axios.post(`https://model-${modelId}.api.baseten.co/production/predict`, data, {
  headers: {
    Authorization: `Api-Key ${BASETEN_API_KEY}`,
  },
})
  .then((response) => {
    const output = response.data.data;
    // Process the output as needed
    console.log('Output:', output);
  })
  .catch((error) => {
    console.error('Error:', error.message || error);
  });
```
This is an example code from https://www.baseten.co/library/stable-diffusion-xl/

The problem is that this is a synchronous call and it's possible that Vercel serverless function would timeout before the model is done.

One of the possible fixes is to ask the prediction service to call a webhook endpoint on my website when the model is done. But webhook feature is not supported by every GPU hosting service out there. For example, Modal.com and baseten.co don't seen to have webhook feature.

So the only option is to keep polling their `get` endpoint on my Vercel server to see whether the model is done. This requires a long running server AFAIK.

How can make a long running server to poll the endpoint? 
If it is impossible because Vercel doesn't support this use case, please suggest another solution. 

Maybe I need a long-running server just to keep polling prediction endpoint for results? If so, is there a service that you suggest?
Thanks!

**Comment by xonx4l** [comment, marked as answer]
Use a queue service like Redis or SQS to offload the long-running task. When your function is called, add a job to the queue. Have a separate serverless function that polls the queue and handles the long-running task.

Use a database like DynamoDB to store state. When your function is called, write a "pending" record to the DB. Have a separate function on a timer that polls the DB for pending records and handles them.

Use a workflow service like Step Functions (AWS) or Temporal to model the async process as a workflow. Your function would start the workflow, which would handle polling and completing the long-running task.

You could run a separate server (e.g. on EC2) that handles polling the prediction API. Have your function write a request to a queue or DB, which the server picks up and processes.

For a service like BaseTen that doesn't support webhooks, you could potentially reverse engineer their API to start a "detached" prediction that you can poll manually. But this may violate their TOS.

The key is to avoid long-running processing in your serverless functions. Queue/DB and separate worker processes are common patterns to handle this. But you likely need some other provider to host the worker logic.
